src/Microsoft.FeatureManagement/ContextualFeatureFilterEvaluator.cs

```diff
@@ -0,0 +1,99 @@
+            {
+                MethodInfo evaluateMethod = targetInterface.GetMethod(nameof(IContextualFeatureFilter<object>.EvaluateAsync), BindingFlags.Public | BindingFlags.Instance);
+
+                _evaluateFunc = TypeAgnosticEvaluate(filter.GetType(), evaluateMethod);
+            }
+
+            _filter = filter;
+        }
+
+        public static Type GetContextualFilterInterface(IFeatureFilterMetadata filter, Type appContextType)
+        {
+            IEnumerable<Type> contextualFilterInterfaces = filter.GetType().GetInterfaces().Where(i => i.IsGenericType && i.GetGenericTypeDefinition().IsAssignableFrom(typeof(IContextualFeatureFilter<>)));
+
+            Type targetInterface = null;
+
+            if (contextualFilterInterfaces != null)
+            {
+                targetInterface = contextualFilterInterfaces.FirstOrDefault(i => i.GetGenericArguments()[0].IsAssignableFrom(appContextType));
+            }
+
+            return targetInterface;
+        }
+
+        public Task<bool> EvaluateAsync(FeatureFilterEvaluationContext evaluationContext, object context)
+        {
+            if (_evaluateFunc == null)
+            {
+                return Task.FromResult(false);
+            }
+
+            return _evaluateFunc(_filter, evaluationContext, context);
+        }
+
+        static Func<object, FeatureFilterEvaluationContext, object, Task<bool>> TypeAgnosticEvaluate(Type filterType, MethodInfo method)
+        {
+            //
+            // Get the generic version of the evaluation helper method
+            MethodInfo genericHelper = typeof(ContextualFeatureFilterEvaluator).GetMethod(nameof(GenericTypeAgnosticEvaluate),
+                BindingFlags.Static | BindingFlags.NonPublic);
+
+            //
+            // Create a type specific version of the evaluation helper method
+            MethodInfo constructedHelper = genericHelper.MakeGenericMethod
+                (filterType, method.GetParameters()[0].ParameterType, method.GetParameters()[1].ParameterType, method.ReturnType);
+
+            //
+            // Invoke the method to get the func
+            object ret = constructedHelper.Invoke(null, new object[] { method });
+
+            return (Func<object, FeatureFilterEvaluationContext, object, Task<bool>>)ret;
+        }
+
+        static Func<object, FeatureFilterEvaluationContext, object, Task<bool>> GenericTypeAgnosticEvaluate<TTarget, TParam1, TParam2, TReturn>(MethodInfo method)
+        {
+            Func<TTarget, FeatureFilterEvaluationContext, TParam2, Task<bool>> func = (Func<TTarget, FeatureFilterEvaluationContext, TParam2, Task<bool>>)Delegate.CreateDelegate
+                (typeof(Func<TTarget, FeatureFilterEvaluationContext, TParam2, Task<bool>>), method);
+
+            Func<object, FeatureFilterEvaluationContext, object, Task<bool>> ret = (object target, FeatureFilterEvaluationContext param1, object param2) => func((TTarget)target, param1, (TParam2)param2);
```

Comment:
>ret [](start = 77, length = 3)

nit: this 'ret' confuse me, how about delegateFunc? Same as line 84.

Reply:
If that is all that confused you then that is awesome :), this code is quite complicated. I'll update it.

---
In reply to: [328399200](https://github.com/microsoft/FeatureManagement-Dotnet/pull/14#discussion_r328399200) [](ancestors = 328399200)